### PR TITLE
feat: handle offline status when starting a call (AR-2145)

### PR DESCRIPTION
**Pull request by gongracr**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/AR-2145" title="AR-2145" target="_blank">SubtaskAR-2145</a>  Calling  - Popup to signal offline therefore functionality is not available
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

When initiating a call while the client was not connected to the internet, we were still launching the `InitiatingCallScreen` which didn't make sense. 

### Solutions

We will be showing a dialog informing about the lack of connectivity every time the user tries to start a call w/o a stable internet connection.

### Dependencies (Optional)

Needs releases with:
https://github.com/wireapp/kalium/pull/827
- [x] GitHub link to other pull request

#### Test Coverage (Optional)

- [x] I have added automated test to this contribution

### Attachments (Optional)
[UPDATED]

https://user-images.githubusercontent.com/2468164/186407938-780764bb-ffbd-4d37-a4c2-30a626e3bf62.mov

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [x] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Comment by AndroidBob** [comment]
Build [1](https://10.10.124.134/job/wire-android-reloaded/job/PR-886/1/) **failed**.

**Comment by ohassine** [comment]
I think we don't have to show the outgoing screen if we don't have an Internet ?

For the string what about using strings from old app ?

Wire 2022-08-23 at 2_52 PM

**Comment by AndroidBob** [comment]
Build [2](https://10.10.124.134/job/wire-android-reloaded/job/PR-886/2/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(4230647)(4230647)-dev-release.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/release/com.wire.android-vv0.0.1(4230647)(4230647)-dev-release.apk)

File: app/src/main/kotlin/com/wire/android/ui/calling/initiating/InitiatingCallScreen.kt [hunk 1]
```diff
@@ -52,15 +58,16 @@ fun InitiatingCallScreen(
...
+    errorState: InitiatingCallState.InitiatingCallError,
+    onErrorStartingCall: () -> Unit,
     callState: CallState,
     toggleMute: () -> Unit,
     toggleSpeaker: () -> Unit,
     toggleVideo: () -> Unit,
     onHangUpCall: () -> Unit,
     onVideoPreviewCreated: (view: View) -> Unit,
-    onSelfClearVideoPreview: () -> Unit
+    onSelfClearVideoPreview: () -> Unit,
```
Comment (ohassine) [inline app/src/main/kotlin/com/wire/android/ui/calling/initiating/InitiatingCallScreen.kt:69]: we don't need this comma  💁‍♂️

File: app/src/main/kotlin/com/wire/android/ui/calling/initiating/InitiatingCallViewModel.kt [hunk 2]
```diff
@@ -116,27 +123,37 @@ class InitiatingCallViewModel @Inject constructor(
...
-            conversationType = conversationType
-        )
-        callRinger.ring(R.raw.ringing_from_me)
+    internal suspend fun initiateCall() {
+        val conversationType = retrieveConversationType()
+        withContext(dispatchers.main()) {
+            val result = startCall(
+                conversationId = conversationId,
+                conversationType = conversationType
+            )
```
Comment (ohassine) [inline app/src/main/kotlin/com/wire/android/ui/calling/initiating/InitiatingCallViewModel.kt:132]: `startCall` should not be run on main

File: app/src/main/kotlin/com/wire/android/ui/calling/initiating/InitiatingCallViewModel.kt [hunk 3]
```diff
@@ -116,27 +123,37 @@ class InitiatingCallViewModel @Inject constructor(
...
-            launch {
-                navigateBack()
-                callRinger.stop()
-            }
-        }
+    fun hangUpCall() = viewModelScope.launch(dispatchers.io()) {
+        endCall(conversationId)
+        callRinger.stop()
+        navigateBack()
+    }
```
Comment (ohassine) [inline app/src/main/kotlin/com/wire/android/ui/calling/initiating/InitiatingCallViewModel.kt:153]: Why did you remove all the coroutines ? We need to run them in parallel..
This way you will see some delay before the screen is closed. `endCall` will take some time before it's done

File: app/src/main/kotlin/com/wire/android/ui/calling/initiating/InitiatingCallViewModel.kt [hunk 4]
```diff
@@ -20,28 +25,31 @@ import com.wire.kalium.logic.feature.call.usecase.ObserveEstablishedCallsUseCase
...
     qualifiedIdMapper: QualifiedIdMapper,
+    private val navigationManager: NavigationManager,
     private val conversationDetails: ObserveConversationDetailsUseCase,
     private val observeEstablishedCalls: ObserveEstablishedCallsUseCase,
     private val startCall: StartCallUseCase,
     private val endCall: EndCallUseCase,
     private val isLastCallClosed: IsLastCallClosedUseCase,
-    private val callRinger: CallRinger
+    private val callRinger: CallRinger,
+    private val dispatchers: DispatcherProvider
```
Comment (ohassine) [inline app/src/main/kotlin/com/wire/android/ui/calling/initiating/InitiatingCallViewModel.kt:48]: I think it's better to change the dispatcher in the suspend function it self and not in the viewModel(eg. usecase). This way we don't have to worry about what Dispatcher to use for what type of work.

**This responsibility lies in the class that does the work.**

**Review by ohassine — CHANGES_REQUESTED** [review]
Seeing the calling UI when we don't have an Internet connection is confusing, it gives me the feeling that the call already started and then the Internet stopped which not the case.

We should display the popup before starting the call I think.

Also in this way, by starting the call we are triggering unnecessary work like initializing AVS and invoking some calling methods

File: app/src/main/kotlin/com/wire/android/ui/calling/initiating/InitiatingCallScreen.kt (hunk 1 above)
Comment (vitorhugods) [reply app/src/main/kotlin/com/wire/android/ui/calling/initiating/InitiatingCallScreen.kt:69]: I think we never agreed on dangling commas, but it's not that bad. It may look a bit uglier now, but it helps in the future when we add more items to the list:

grafik

Credits: https://medium.com/@nikgraf/why-you-should-enforce-dangling-commas-for-multiline-statements-d034c98e36f8

File: app/src/main/kotlin/com/wire/android/ui/calling/initiating/InitiatingCallViewModel.kt (hunk 2 above)
Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/calling/initiating/InitiatingCallViewModel.kt:132]: io? or the default thread then?

File: app/src/main/kotlin/com/wire/android/ui/calling/initiating/InitiatingCallViewModel.kt (hunk 3 above)
Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/calling/initiating/InitiatingCallViewModel.kt:153]: you were launching 3 threads here, 1 containing another 2. Could we do it then with just 1 extra one to end the call and leave the current onging thread to run the tasks to stop the ring and navigate back?

File: app/src/main/kotlin/com/wire/android/ui/calling/initiating/InitiatingCallViewModel.kt (hunk 4 above)
Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/calling/initiating/InitiatingCallViewModel.kt:44]: Hmmm not sure if I follow. I do use the dispatchers to change the context on the corresponding method inside the view model. 

If you mean that the suspend function calling the view model should hold the responsibility (`InitiateCallScreen` for instance) you are implying that the view layer knows already the kind of logic that is being performed on the VM (whether if it is a IO operation or not) which I'm not sure it is the case.

**Comment by AndroidBob** [comment]
Build [4](https://10.10.124.134/job/wire-android-reloaded/job/PR-886/4/) **failed**.

File: app/src/main/kotlin/com/wire/android/ui/calling/initiating/InitiatingCallViewModel.kt (hunk 3 above)
Comment (ohassine) [reply app/src/main/kotlin/com/wire/android/ui/calling/initiating/InitiatingCallViewModel.kt]: I am lunching 3 coroutines 1 parent and 2 children. This will not affect the app performance anymore since coroutines are cheap compared to threads. The most important thing here is that they should run in parallel

File: app/src/main/kotlin/com/wire/android/ui/calling/initiating/InitiatingCallViewModel.kt (hunk 4 above)
Comment (ohassine) [reply app/src/main/kotlin/com/wire/android/ui/calling/initiating/InitiatingCallViewModel.kt:44]: I mean we don't need to change dispatcher here, we should do it in the suspend function itself(usecase for example)
All that we need to do here is starting a new coroutine and call the functions

Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/calling/initiating/InitiatingCallViewModel.kt]: addressed

**Comment by AndroidBob** [comment]
Build [6](https://10.10.124.134/job/wire-android-reloaded/job/PR-886/6/) **failed**.

**Comment by AndroidBob** [comment]
Build [8](https://10.10.124.134/job/wire-android-reloaded/job/PR-886/8/) **failed**.

**Comment by AndroidBob** [comment]
Build [9](https://10.10.124.134/job/wire-android-reloaded/job/PR-886/9/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(4244435)(4244435)-dev-release.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/release/com.wire.android-vv0.0.1(4244435)(4244435)-dev-release.apk)
